File: examples/python/03_capture_equality.py
import json

import oc_sidememory

from _transformers import accepts, load_runtime, masked_next_token


model, tokenizer, vocabulary = load_runtime()
schema = json.dumps({
    "type": "object",
    "properties": {"id": {"type": "number"}, "confirmation": {"type": "number"}},
    "required": ["id", "confirmation"],
    "additionalProperties": False,
})
extensions = json.dumps({
    "version": 1,
    "objects": [{
        "schemaPath": "$",
        "propertyOrder": ["id", "confirmation"],
        "captures": [{"name": "primary", "sourceProperty": "id"}],
        "relations": [{
            "targetProperty": "confirmation",
            "operator": "equal",
            "capture": "primary",
        }],
    }],
})
compiled = oc_sidememory.compile_schema(
    schema, vocabulary, tokenizer.vocab_size, extensions_json=extensions
)
guide = oc_sidememory.SidememoryGuide(compiled)
assert accepts(guide, tokenizer, '{"id":1,"confirmation":1.0}')
assert not accepts(guide, tokenizer, '{"id":1,"confirmation":2}')
selected = masked_next_token(model, tokenizer, guide)
print(f"capture equality: real-model token {selected!r}; mismatch rejected")

File: examples/rust/03_capture_equality.rs
mod common;

use oc_sidememory::sidememory::ProbeDecision;

fn main() {
    let schema = r#"{"type":"object","properties":{"id":{"type":"number"},"confirmation":{"type":"number"}},"required":["id","confirmation"],"additionalProperties":false}"#;
    let extensions = r#"{"version":1,"objects":[{"schemaPath":"$","propertyOrder":["id","confirmation"],"captures":[{"name":"primary","sourceProperty":"id"}],"relations":[{"targetProperty":"confirmation","operator":"equal","capture":"primary"}]}]}"#;
    let compiled = common::compile(
        schema,
        &[br#"{"id":1,"confirmation":1.0}"#, br#"{"id":1,"confirmation":2}"#],
        Some(extensions),
    );
    let mut guide = common::guide(compiled);
    assert_eq!(guide.probe(0).expect("probe succeeds"), ProbeDecision::Allow);
    assert!(matches!(guide.probe(1).expect("probe succeeds"), ProbeDecision::Deny(_)));
    println!("capture equality: accepted 1 == 1.0 and rejected 1 != 2");
}
